### PR TITLE
[RAISE-BP] Add support for tt.broadcast increasing tensor rank

### PR DESCRIPTION
**Pull request by mfrancepillois**
Extend the broadcast support to handle `tt.broadcast` operations when the rank of the destination tensor is greater than the rank of the source tensor.

Closes Issue: #1429

### CODE REVIEW

**Review by victor-eds — COMMENTED** [review]
Can we have a test with:
```
%1 = tt.broadcast %cst : tensor<128xi32> -> tensor<256x2x128xi32>
```
? Incrementing the rank in 2

File: third_party/intel/lib/TritonRaiseBlockPointer/TritonRaiseBlockPointer.cpp [hunk 1]
```diff
@@ -574,15 +574,62 @@ TritonRaiseBlockPointer::visitAddPointerOperand(triton::BroadcastOp broadcastOp,
...
   if (failed(visitOperand(src, state, loc, builder))) {
     return failure();
   }
 
-  llvm::copy(dstShape, state.sizes.begin());
+  if (srcShape.size() == dstShape.size()) {
+    llvm::copy(dstShape, state.sizes.begin());
+  } else {
+    // Offset must be equal, otherwise we don.t know which offset should be
+    // propagate to the new axis.
```
Comment (etiotto) [inline third_party/intel/lib/TritonRaiseBlockPointer/TritonRaiseBlockPointer.cpp:588]: typo: propagate -> propagated

File: third_party/intel/lib/TritonRaiseBlockPointer/TritonRaiseBlockPointer.cpp [hunk 2]
```diff
@@ -574,15 +574,62 @@ TritonRaiseBlockPointer::visitAddPointerOperand(triton::BroadcastOp broadcastOp,
...
+    for (int i = 1; i < state.offsets.size(); i++) {
```
Comment (etiotto) [inline third_party/intel/lib/TritonRaiseBlockPointer/TritonRaiseBlockPointer.cpp:589]: [nit]: use ++i (also for the other loop in the code below)

File: third_party/intel/lib/TritonRaiseBlockPointer/TritonRaiseBlockPointer.cpp [hunk 3]
```diff
@@ -574,15 +574,62 @@ TritonRaiseBlockPointer::visitAddPointerOperand(triton::BroadcastOp broadcastOp,
...
+
+    // Create the new axis.
+    // The positions of the new axis are determined based and the shape values.
+    // If shape are the same, the new axis are added at the end.
+    size_t srcAxis = 0;
+    for (size_t axis = 0; axis < dstShape.size(); axis++) {
+      if ((srcAxis < srcShape.size()) &&
+          (srcShape[srcAxis] == dstShape[axis])) {
+        srcAxis++;
+      } else {
```
Comment (etiotto) [inline third_party/intel/lib/TritonRaiseBlockPointer/TritonRaiseBlockPointer.cpp:606]: Use early "continue":
if (condition) {
   srcAxis++;
   continue;
 }
 
 <else part here>

**Review by etiotto — APPROVED** [review]
LGTM - modulo the code review comment left inline

**Review by etiotto — APPROVED** [review]

**Comment by etiotto** [comment]
@victor-eds @whitneywhtsang are you OK with the PR?

**Review by whitneywhtsang — APPROVED** [review]

**Review by victor-eds — APPROVED** [review]

**Pull request merged**